File: packages/compose/nockerl-design-components/src/main/kotlin/com/dizyx/nockerl/design/components/NockerlChip.kt
package com.dizyx.nockerl.design.components

import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Close
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.unit.dp
import com.dizyx.nockerl.design.tokens.LocalNockerlColors
import com.dizyx.nockerl.design.tokens.NockerlPillShape

/**
 * The standardized **chip**: a filled keycap **pill** (the session-chip /
 * filter-chip idiom). One of the two surfaces that deliberately keep the ~50%
 * [NockerlPillShape] (the other is the input bar); everything else uses the
 * 12dp control radius. This contrast (pills for chips + input, the control
 * radius for buttons) is the point Design Review #1 locked in.
 *
 * Two states:
 * - **selected**: a solid accentPrimary fill with a contrast-picked label (the
 *   active session keycap).
 * - **unselected**: a soft accentPrimarySoft fill with a cyan label, so the row
 *   reads as a cohesive chip strip while the selected one dominates.
 *
 * Color flows entirely from the palette; the shape flows from [NockerlPillShape].
 *
 * @param text the chip label.
 * @param onClick invoked on tap.
 * @param modifier outer modifier.
 * @param selected whether this chip is the active selection.
 * @param enabled when `false`, the chip is dimmed and non-interactive.
 * @param contentPadding inner padding (defaults to a compact keycap padding).
 * @param leadingContent optional leading content (e.g. a status dot) before the label.
 * @param onRemove when non-null, the chip is REMOVABLE: a trailing ✕ renders with
 *   its own tap target (the ratified chip contract is the capability UNION
 *   of selectable, removable, and disabled, on every platform).
 */
@Composable
fun NockerlChip(
    text: String,
    onClick: () -> Unit,
    modifier: Modifier = Modifier,
    selected: Boolean = false,
    enabled: Boolean = true,
    contentPadding: PaddingValues = PaddingValues(horizontal = 14.dp, vertical = 8.dp),
    leadingContent: (@Composable () -> Unit)? = null,
    onRemove: (() -> Unit)? = null,
) {
    val colors = LocalNockerlColors.current
    val fill = if (selected) colors.accentPrimary else colors.accentPrimarySoft
    val content = if (selected) pickOnAccent(colors.accentPrimary) else colors.accentPrimary
    val resolvedFill = if (enabled) fill else fill.copy(alpha = fill.alpha * DISABLED_ALPHA)
    val resolvedContent = if (enabled) content else content.copy(alpha = DISABLED_ALPHA)

    NockerlControlContainer(
        onClick = onClick,
        enabled = enabled,
        fill = resolvedFill,
        shape = NockerlPillShape,
        // Ratified: the SELECTED solid-accent keycap wears the top catch-light.
        litHighlight = selected && enabled,
        modifier = modifier,
    ) {
        Row(
            modifier = Modifier.padding(contentPadding),
            horizontalArrangement = Arrangement.spacedBy(6.dp, Alignment.CenterHorizontally),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            leadingContent?.invoke()
            Text(
                text = text,
                style = NockerlControlTextStyle,
                color = resolvedContent,
            )
            if (onRemove != null) {
                Icon(
                    imageVector = Icons.Filled.Close,
                    contentDescription = "Remove $text",
                    tint = resolvedContent,
                    modifier =
                        Modifier
                            .size(16.dp)
                            .clickable(enabled = enabled, onClick = onRemove),
                )
            }
        }
    }
}

File: Sources/NockerlDesign/Components/NockerlChip.swift
// NockerlChip: the standardized chip, a filled keycap PILL (the session-chip / filter-chip
// idiom). One of the two surfaces that deliberately keep the fully-rounded pill silhouette
// (the other is the input bar); every other control uses the 12pt control radius. This
// contrast (pills for chips + input, control radius for buttons) is the point Design
// Review #1 locked in (design-laws §4).
//
// The ratified cyan grammar + the full capability UNION (one contract:
// selectable + removable + disabled on every platform):
//   - selected:    SOLID `accentPrimary` fill + contrast-picked label + the shared
//                  top catch-light (never a gradient).
//   - unselected:  `accentPrimarySoft` wash + `accentPrimary` label.
//   - removable:   an optional trailing ✕ with its own tap.
//
// Same platform-honoring feedback as the rest of the family (design-laws §7); tokens-pure.

import SwiftUI

/// A single token-driven chip, a filled keycap pill. `selected` toggles the accent
/// selection wash; feedback matches the button family (static fill, animated
/// opacity/brightness/scale).
public struct NockerlChip: View {
    private let text: String
    private let selected: Bool
    private let action: () -> Void
    private let onRemove: (() -> Void)?

    /// Create a chip.
    /// - Parameters:
    ///   - text: the chip label.
    ///   - selected: whether this chip is the active selection (defaults to `false`).
    ///   - action: invoked on tap.
    ///   - onRemove: when non-nil, the chip is REMOVABLE. A trailing ✕ renders with
    ///     its own tap (ruling C6: the chip contract is the capability union).
    public init(
        _ text: String,
        selected: Bool = false,
        action: @escaping () -> Void,
        onRemove: (() -> Void)? = nil
    ) {
        self.text = text
        self.selected = selected
        self.action = action
        self.onRemove = onRemove
    }

    public var body: some View {
        Button(action: action) {
            HStack(spacing: NockerlSpace.space1) {
                Text(text)
                if let onRemove {
                    // The innermost interactive view wins the tap, so the remove
                    // affordance rides inside the chip's own label.
                    Button(action: onRemove) {
                        Image(systemName: "xmark")
                            .font(.system(size: NockerlFontSize.size10, weight: .medium))
                    }
                    .buttonStyle(.plain)
                    .accessibilityLabel("Remove \(text)")
                }
            }
        }
        .buttonStyle(ChipStyle(selected: selected))
    }
}

/// The `ButtonStyle` behind ``NockerlChip``. Reads the environment via a private nested view
/// (the idiomatic path for `colorScheme` / `isEnabled` inside a `ButtonStyle`).
private struct ChipStyle: ButtonStyle {
    let selected: Bool

    static let disabledAlpha: Double = 0.38

    func makeBody(configuration: Configuration) -> some View {
        // Not named `Body`: a nested `Body` binds as the `ButtonStyle.Body` associatedtype
        // witness and leaks a visibility requirement. Opaque return keeps it inferred.
        StyleBody(selected: selected, configuration: configuration)
    }

    private struct StyleBody: View {
        let selected: Bool
        let configuration: Configuration

        @Environment(\.colorScheme) private var colorScheme
        @Environment(\.isEnabled) private var isEnabled

        var body: some View {
            let palette = NockerlPalette.resolve(colorScheme)
            // Ratified B2: selected → SOLID accent + contrast-picked ink + catch-light;
            // unselected → the soft cyan wash + cyan ink (the cohesive chip strip).
            // The fill is static; only feedback props animate.
            let fill: Color = selected ? palette.accentPrimary : palette.accentPrimarySoft
            let label: Color =
                selected ? NockerlContrast.pickOn(palette.accentPrimary) : palette.accentPrimary

            configuration.label
                .font(.nockerl(size: NockerlFontSize.size12, weight: .medium))
                .foregroundColor(label)
                .padding(.horizontal, NockerlSpace.space3)
                .padding(.vertical, NockerlSpace.space2)
                .background(fill)
                .overlay(alignment: .top) {
                    if selected, isEnabled {
                        // RULING: one identical catch-light everywhere, the SAME
                        // surfaceHighlight token the web binds (--color-surface-highlight).
                        Rectangle()
                            .fill(palette.surfaceHighlight)
                            .frame(height: NockerlSpace.spacePx)
                    }
                }
                .clipShape(RoundedRectangle(cornerRadius: NockerlRadius.pill, style: .continuous))
                .contentShape(RoundedRectangle(cornerRadius: NockerlRadius.pill, style: .continuous))
                .opacity(isEnabled ? 1 : ChipStyle.disabledAlpha)
                .nockerlPressFeedback(isPressed: configuration.isPressed)
        }
    }
}
